### PR TITLE
Strings as Categorical Values

### PR DESCRIPTION
**Pull request by yanchaomei**
The decision tree of lightgbm can only support the classification of integer labels. For string type data labels, users need to preprocess the labels themselves, convert the string type to integer type, and then use lightgbm.

We modify and support the direct input of string type data, establish a string to double mapping when reading the file in the training process, and then save it to the model. Restore the map from the model during the prediction process;

### CODE REVIEW

File: include/LightGBM/dataset.h [hunk 1]
```diff
@@ -13,6 +13,8 @@
 #include <LightGBM/utils/random.h>
 #include <LightGBM/utils/text_reader.h>
 
+// #include "../../src/io/parser.hpp"
```
Comment (hippogr) [inline include/LightGBM/dataset.h:16]: >/ #include "../../src/io/parser.hpp" [](http://example.com/codeflow?start=1&length=36)

if you don't need this line, you can simply remove it instead of commenting it. :-)

File: include/LightGBM/dataset.h [hunk 2]
```diff
@@ -739,6 +754,36 @@ class Dataset {
...
+        name = name + iter->first + "=" + std::to_string(iter->second) + '\n'; 
+        iter++;
+      }
+      name += '\n';
+      mapping.push_back(name);
+    }
+  }
+
+
+  inline std::vector< std::string >  get_mapping() const {
```
Comment (hippogr) [inline include/LightGBM/dataset.h:783]: > std::string  [](http://example.com/codeflow?start=21&length=13)

Looks like that we don't need the space between std::string

File: include/LightGBM/dataset.h [hunk 3]
```diff
@@ -24,6 +26,19 @@
 
 namespace LightGBM {
 
+class Str2Num {
```
Comment (hippogr) [inline include/LightGBM/dataset.h:29]: >class [](http://example.com/codeflow?start=0&length=5)

since all of the members are public, so how about change it to structure instead of class?

File: include/LightGBM/dataset.h [hunk 4]
```diff
@@ -781,6 +826,7 @@ class Dataset {
   std::vector<int> numeric_feature_map_;
   int num_numeric_features_;
   std::string parser_config_str_;
+  std::vector< std::string > mapping;
```
Comment (hippogr) [inline include/LightGBM/dataset.h:829]: >< std::string  [](http://example.com/codeflow?start=13&length=14)

same

File: include/LightGBM/dataset.h [hunk 5]
```diff
@@ -739,6 +754,36 @@ class Dataset {
     return raw_data_[numeric_feature_map_[feat_ind]].data();
   }
 
+  inline void store_mapping()  {
```
Comment (hippogr) [inline include/LightGBM/dataset.h:757]: >   [](http://example.com/codeflow?start=29&length=2)

single space is OK

File: include/LightGBM/dataset.h [hunk 6]
```diff
@@ -739,6 +754,36 @@ class Dataset {
...
+    // store maps to mapping
+    int offset = 0;
+    for(int i=0; i < maps.size(); i++){
```
Comment (hippogr) [inline include/LightGBM/dataset.h:760]: >for(i [](http://example.com/codeflow?start=4&length=5)

for<space>{

File: include/LightGBM/dataset.h [hunk 7]
```diff
@@ -739,6 +754,36 @@ class Dataset {
     return raw_data_[numeric_feature_map_[feat_ind]].data();
   }
 
+  inline void store_mapping()  {
+    // store maps to mapping
+    int offset = 0;
+    for(int i=0; i < maps.size(); i++){    
+      std::string name;
+      
+      if(i == label_idx_){
```
Comment (hippogr) [inline include/LightGBM/dataset.h:763]: > if( [](http://example.com/codeflow?start=5&length=4)

same one, please review the other code to make sure you are following the code style

File: include/LightGBM/dataset.h [hunk 8]
```diff
@@ -739,6 +754,36 @@ class Dataset {
     return raw_data_[numeric_feature_map_[feat_ind]].data();
   }
 
+  inline void store_mapping()  {
+    // store maps to mapping
+    int offset = 0;
+    for(int i=0; i < maps.size(); i++){    
+      std::string name;
```
Comment (hippogr) [inline include/LightGBM/dataset.h:761]: >std::string name; [](http://example.com/codeflow?start=6&length=17)

how about move this one out of loop?

File: include/LightGBM/dataset.h [hunk 9]
```diff
@@ -739,6 +754,36 @@ class Dataset {
...
+        name = "label";
+        offset--;
+      }else{ 
+        name = feature_names_[i + offset];
+        }
+        
+      name = name + '\n';
+      std::map<std::string, double>::iterator iter;
+      iter = maps[i].Str2NumMap.begin();
+      while(iter != maps[i].Str2NumMap.end()){
```
Comment (hippogr) [inline include/LightGBM/dataset.h:773]: why not 
`for (auto i = maps[i].Str2NumMap..begin(); i != maps[i].Str2NumMap..end(); ++i)`?

File: include/LightGBM/dataset.h [hunk 10]
```diff
@@ -739,6 +754,36 @@ class Dataset {
...
+        name = name + iter->first + "=" + std::to_string(iter->second) + '\n'; 
+        iter++;
+      }
+      name += '\n';
+      mapping.push_back(name);
```
Comment (hippogr) [inline include/LightGBM/dataset.h:778]: >push_back(name) [](http://example.com/codeflow?start=14&length=15)

push_back(std::move(name)) is better. You can check the other places to see if you can use move to avoid additional costs.

File: include/LightGBM/utils/common.h [hunk 11]
```diff
@@ -31,6 +31,7 @@
 #include <unordered_map>
 #include <utility>
 #include <vector>
+#include <map>
```
Comment (hippogr) [inline include/LightGBM/utils/common.h:34]: is this the only line to change it in this file? I suggest that you add this include in the cpp or header when it uses <map> instead of adding it in the common header

File: include/LightGBM/dataset.h [hunk 12]
```diff
@@ -739,6 +754,36 @@ class Dataset {
...
+      std::string name;
+      
+      if(i == label_idx_){
+        name = "label";
+        offset--;
+      }else{ 
+        name = feature_names_[i + offset];
+        }
+        
+      name = name + '\n';
```
Comment (hippogr) [inline include/LightGBM/dataset.h:770]: >name = name + '\n'; [](http://example.com/codeflow?start=6&length=19)

I see you are using += when you process string, which is good. But why not this one? :-)

File: src/boosting/gbdt_model_text.cpp [hunk 13]
```diff
@@ -621,6 +635,35 @@ bool GBDT::LoadModelFromString(const char* buffer, size_t len) {
...
+        is_inmaps = true;
+      } else if (cur_line == std::string("end of Maps")) {
+        p += line_len;
+        p = Common::SkipNewLine(p);
+        break;
+      } else if(cur_line == std::string("map")){
+        index++;
+        tmp.Str2NumMap.clear();
+      }
+       else if (is_inmaps) {
```
Comment (hippogr) [inline src/boosting/gbdt_model_text.cpp:655]: > else [](http://example.com/codeflow?start=6&length=5)

keep the same style as the code above

File: src/boosting/gbdt_model_text.cpp [hunk 14]
```diff
@@ -621,6 +635,35 @@ bool GBDT::LoadModelFromString(const char* buffer, size_t len) {
...
+         std::vector< std::string > key_value;
+         key_value = Common::Split(cur_line.c_str(), "=");
+         
+         tmp.Str2NumMap.insert(std::pair < std::string , double > (key_value[0], std::stod(key_value[1]) ) );
```
Comment (hippogr) [inline src/boosting/gbdt_model_text.cpp:659]: >) ) ) [](http://example.com/codeflow?start=103&length=5)

same question about the style

File: src/boosting/gbdt_model_text.cpp [hunk 15]
```diff
@@ -621,6 +635,35 @@ bool GBDT::LoadModelFromString(const char* buffer, size_t len) {
...
+        p += line_len;
+        p = Common::SkipNewLine(p);
+        break;
+      } else if(cur_line == std::string("map")){
+        index++;
+        tmp.Str2NumMap.clear();
+      }
+       else if (is_inmaps) {
+         std::vector< std::string > key_value;
+         key_value = Common::Split(cur_line.c_str(), "=");
```
Comment (hippogr) [inline src/boosting/gbdt_model_text.cpp:657]: I guess here you need some checks. One is that you may need to check if the key_value vector has two elements so your accessing key_value[1] won't bring exceptions. Another one is to do check if std::stod throws exception or not.

File: src/io/dataset_loader.cpp [hunk 16]
```diff
@@ -17,6 +18,8 @@ namespace LightGBM {
 
 using json11::Json;
 
+extern std::vector<Str2Num> maps;
```
Comment (hippogr) [inline src/io/dataset_loader.cpp:21]: why do you need this?

File: src/boosting/gbdt_model_text.cpp [hunk 17]
```diff
@@ -16,6 +17,8 @@
 
 namespace LightGBM {
 
+extern std::vector<Str2Num> maps;
+
```
Comment (hippogr) [inline src/boosting/gbdt_model_text.cpp:21]: you don't really need this, right?

File: src/io/parser.cpp [hunk 18]
```diff
@@ -12,6 +12,8 @@
 
 namespace LightGBM {
 
+std::vector<Str2Num> maps;
```
Comment (hippogr) [inline src/io/parser.cpp:15]: same, seems like you don't need this

File: src/io/parser.hpp [hunk 19]
```diff
@@ -15,6 +15,135 @@
 
 namespace LightGBM {
 
+// define a str2num map class
+
+extern std::vector<Str2Num> maps;
+
+const char* Atof_and_map(const char* p, double* out, int idx) {
```
Comment (hippogr) [inline src/io/parser.hpp:22]: >Atof_and_map [](http://example.com/codeflow?start=12&length=12)

I guess you don't need to write a string to float function by yourself. How about read out the string, trim the spaces and tabs, then check if it is a nan or infinite, then you can use the API to do that. I don't see any necessary to implement a str2num here

File: tests/python_package_test/test_string.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+import numbers
+from pathlib import Path
+
+import numpy as np
+# import pytest
+from scipy import sparse
+from sklearn.datasets import dump_svmlight_file, load_svmlight_file, load_breast_cancer
+from sklearn.model_selection import train_test_split
+import sys
+sys.path.append("python-package")
```
Comment (jameslamb) [inline tests/python_package_test/test_string.py:12]: Please do not modify `sys.path`, as that could lead to a situation where some imports seem to work during tests and then actually fail when users run `lightgbm` on their systems.

File: tests/python_package_test/test_string.py [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+        "pred_early_stop_freq": 5,
+        "pred_early_stop_margin": 1.5
+    }
+    pred_early_stopping = bst.predict(X_test, **pred_parameter)
+    # scores likely to be different, but prediction should still be the same
+    np.testing.assert_array_equal(np.sign(pred_from_matr),
+                                  np.sign(pred_early_stopping))
+
+
+test_basic(Path(r'tests\python_package_test\test_string'))
```
Comment (jameslamb) [inline tests/python_package_test/test_string.py]:
```suggestion
```

It is not necessary to call test functions like this. `pytest` (the test runner used in LightGBM), will automatically detected and run all functions whose names start with `test_`.

Please remove this line, and ensure that every file has a newline at the end of it.

File: src/c_api.cpp [hunk 22]
```diff
@@ -728,6 +728,7 @@ class Booster {
 
   std::string SaveModelToString(int start_iteration, int num_iteration,
                                 int feature_importance_type) const {
+
```
Comment (jameslamb) [inline src/c_api.cpp:731]: Please revert this and any other whitespace-only changes not related to this PR. Such changes increase the size of the diff and therefore the effort required to review.

File: src/io/dataset_loader.cpp [hunk 23]
```diff
@@ -1172,7 +1175,7 @@ void DatasetLoader::ConstructBinMappersFromTextData(int rank, int num_machines,
   if (dataset->has_raw()) {
     dataset->ResizeRaw(static_cast<int>(sample_data.size()));
   }
-
+
```
Comment (jameslamb) [inline src/io/dataset_loader.cpp:1178]:
```suggestion
```

File: src/io/dataset_loader.cpp [hunk 24]
```diff
@@ -1364,6 +1368,7 @@ void DatasetLoader::ExtractFeaturesFromFile(const char* filename, const Parser*
       OMP_LOOP_EX_END();
     }
     OMP_THROW_EX();
+
```
Comment (jameslamb) [inline src/io/dataset_loader.cpp:1371]:
```suggestion
```

File: tests/cpp_tests/test.conf [hunk 25]
```diff
@@ -0,0 +1,171 @@
+# task type, support train and predict
+task = train
+
+
```
Comment (jameslamb) [inline tests/cpp_tests/test.conf:4]:
```suggestion
```

Please avoid using more than one consecutive blank line in these `.conf` and `.config` files, to make them easier to read.

**Review by jameslamb — CHANGES_REQUESTED** [review]
This PR currently is proposing adding several files to source control for this project which I don't approve of adding to this repo. Please remove them or explain why you believe they should be added.

* `LightGBM.pyproj`
* `CMakeSettings.json`
* `windows/LightGBM_model.txt`
* `windows/LightGBM_predict_result.txt`

I've also left several other comments. Please don't include whitespace-only changes or commented-out code in this pull request.

**Comment by jameslamb** [comment]
Given that there hasn't been any activity on this PR in 3+ months and the CLA has not been signed, it seems to me that this PR has been abandoned. I'm going to close this.

If anyone reading this is interested in picking up this work for #789, please open a new pull request.
